server/src/main/java/org/opensearch/index/analysis/AnalysisRegistry.java

```diff
@@ -221,6 +222,51 @@ public Analyzer getAnalyzer(String analyzer) throws IOException {
         return analyzerProvider.get(environment, analyzer).get();
     }
 
+    /**
+     * Returns registered keys of {@link Analyzer}s on this node.
+     * TODO: Verify in the test that returned {@link Set} is unmodifiable, ie. throws {@link UnsupportedOperationException}
```

Comment:
The TODO item is redundant I believe: `AnalysisRegistry` constructor initializes all components as `unmodifiableMap` which makes keys and values unmodifiable as well.

Reply:
Yes, they are constructed as `unmodifiableMap` in the ctor but it is not documented and I think it is not tested as well.
The problem here is that registers of analysis components (like analyzers, tokenizers, ... etc) have never been exposed like this before and by exposing a set of keys (via `this.analyzers.keySet()`) we can put the internal analyzers registry into risk as well, at least this is how I read [`.keySet()` JavaDoc](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--).

If the ctor is changed in the future then without explicit test that returned {@link Set} is unmodifiable we will open a way for any plugin to manipulate analysis components. That is why I would like to be quite defensive and provide these tests.

Reply:
Very fair and I think it brings a good point that this API is actually not what we want at all. What we may want it may something along this lines:

```
class AnalysisComponent {
    private Class<?> component;
    private String> name;
    boolean preconfigured;
}

public List<AnalysisComponent> getAnalysisComponents() {
  ...
}

```

This is generic API which:
a) does not provide any exposure to internals of `AnalysisRegistry`
b) provides fairly generic registry introspection without revealing what is actually could (and may in the future) manage

Reply:
@lukas-vlcek do @reta's points here make sense?

Reply:
> a) does not provide any exposure to internals of AnalysisRegistry

Agree. Good point!

> b) provides fairly generic registry introspection without revealing what AnalysisRegistry actually could (and may in the future) manage

I am not sure if we can get anything more than just the pure list of keys for individual analysis components. For example for analyzers we can get:
```json
{  "analyzers" : [
    "arabic", "armenian", "basque", "bengali", "brazilian", "bulgarian",
    "catalan", "chinese", "cjk", "czech", "danish", "dutch", "english", "estonian",
    ...
    "russian", "snowball" "sorani", "spanish", "swedish", "thai", "turkish"
    ]
}
```
and that's it. I do not think we can get more detailed information. The reason is that unless the analysis component is used it may not be instantiated (ie. `AnalysisModule.getAnalysisRegistry().getAnalyzers()`).

So instead of returning List of some objects I think we can only return `List<String>`. And what I am going to change is that this list will not be connected to internal analysis register Maps (so changes in returned List will not impact internal registers).

WDYT?

Reply:
> I am not sure if we can get anything more than just the pure list of keys for individual analysis components. For example for analyzers we can get:

The one example of meta information that included into the example is `preconfigured` flag (prebuilt components), this is not covered by current implementation but probably should

Reply:
@reta FYI, for now all those get_X_Keys() methods return defensive registry key `Set` copy.
Not great but perhaps the easiest solution ATM – see my later https://github.com/opensearch-project/OpenSearch/pull/10296#issuecomment-1883373794

Reply:
@lukas-vlcek I have concerns related to APIs (which is in my opinion is too low level and addressed the immediate need but does not include long time vision), perhaps I am overthinking this part so I would let other maintainers to share their views.

Reply:
@reta Feel free to share any kind of long term vision around analysis components.

The original idea was to be able to pull list of names of available analysis components to address lacking documentation and give users option to get run-time names list. I think that getting more detailed information about  individual components is challenging because they are a) initialized lazily and b) there does not seem to be any API that can reveal internal details (for example which token filters are used in specific analyzer, and this goes all the way down to Lucene level).

Reply:
> @reta Feel free to share any kind of long term vision around analysis components.

I already did share suggestion for more lean APIs

Reply:
@reta I tried to respond here https://github.com/opensearch-project/OpenSearch/pull/10296#issuecomment-1883373794